#!/usr/bin/env snakemake

# ───────────────────────────────────────────────
# IMPORT 
# ───────────────────────────────────────────────

import pandas as pd
import os
import sys

from collections import defaultdict

# ───────────────────────────────────────────────
# CONFIG 
# ───────────────────────────────────────────────

# Define minimum snakemake version
min_snakemake_version = "8.25.0"

# Set snakemake main workdir variable. 
SNAKEDIR = workflow.basedir + '/'

# Load configuration from config file
configfile: SNAKEDIR + "../config/config.yaml"


# ───────────────────────────────────────────────
# LOCAL VARIABLE DEFINITIONS 
# ───────────────────────────────────────────────

# Utility function to make sure that values we expect to be booleans are passed as such
def as_bool(val):
    if isinstance(val, bool):
        return val
    if isinstance(val, str):
        return val.strip().lower() in {"true", "yes", "1"}
    return False


# ───────────────────────────────────────────────
# Define Output Directory Structure
# ───────────────────────────────────────────────

PREFIX = config["prefix"]

# Transdecoder outputs
TRANSDECODER_OUT_DIR = "02_ORF_prediction/transdecoder"
TRANSDECODER_LOG_DIR = f"logs/{TRANSDECODER_OUT_DIR}"

# GMST outputs
GMST_OUT_DIR = "02_ORF_prediction/gmst"
GMST_LOG_DIR = f"logs/{GMST_OUT_DIR}"

# Annotate output folders
ANNOTATE_DIR = "03_annotate"
ANNOTATE_SUBDIRS = {
    "annot":       f"{ANNOTATE_DIR}/1_annotation",
    "reclocus":    f"{ANNOTATE_DIR}/2_reclocus",
    "cds":         f"{ANNOTATE_DIR}/3_cds_integration",
    "asef":        f"{ANNOTATE_DIR}/4_ASEF",
    "poison_exon": f"{ANNOTATE_DIR}/5_poison_exons",
    "final":       f"{ANNOTATE_DIR}/6_final",
}

# Pfamscan output folders
PFAMSCAN_OUT_DIR = "04_ORF_protein_domains/pfamscan"
PFAMSCAN_LOG_DIR = f"logs/{PFAMSCAN_OUT_DIR}"

# Interproscan output folders
INTERPROSCAN_OUT_DIR = "04_ORF_protein_domains/interproscan"
INTERPROSCAN_LOG_DIR = f"logs/{INTERPROSCAN_OUT_DIR}"

# ───────────────────────────────────────────────
# RULES
# ───────────────────────────────────────────────

include: "rules/01_isoform-counts.smk"
include: "rules/02_ORF_prediction_cpatv3.smk"
include: "rules/02_ORF_prediction_transdecoder.smk"
include: "rules/02_ORF_prediction_gmst.smk"
include: "rules/03_isoPropeller_annotate.smk"
include: "rules/04_ORF_protein_domains_interproscan.smk"
include: "rules/04_ORF_protein_domains_pfamscan.smk"
include: "rules/05_genomic_element_overlaps.smk"
include: "rules/06_tracks.smk"
include: "rules/07_splicechains.smk"
include: "rules/08_DEG.smk"


rule all:
    input:
        # --- Step 01: Isoform Counts ---
        f"01_isoform_counts/{PREFIX}_isoform-counts.txt",

        # --- Step 02: ORF Prediction ---
        # CPAT outputs
        f"02_ORF_prediction/cpat_leng/{PREFIX}_corrected.cpatv3l18.ORF_seqs.fa",
        # TransDecoder outputs
        f"{TRANSDECODER_OUT_DIR}/merged/{PREFIX}.transdecoder.pep",
        f"{TRANSDECODER_OUT_DIR}/merged/{PREFIX}.transdecoder.cds",
        f"{TRANSDECODER_OUT_DIR}/merged/{PREFIX}.transdecoder.gff3",
        f"{TRANSDECODER_OUT_DIR}/merged/{PREFIX}.transdecoder.bed",
        # GMST outputs
        f"{GMST_OUT_DIR}/merged/{PREFIX}.gmst.faa",
        f"{GMST_OUT_DIR}/merged/{PREFIX}.gmst.fnn",
        f"{GMST_OUT_DIR}/merged/{PREFIX}.gmst.gff3",

        # --- Step 03: Annotation & Analysis ---
        f"{ANNOTATE_SUBDIRS['final']}/{PREFIX}_reference_reclocus_refined.txt",
        f"{ANNOTATE_SUBDIRS['final']}/{PREFIX}_reference_reclocus_CDS_aa.fa",
        f"{ANNOTATE_SUBDIRS['asef']}/{PREFIX}_reference_reclocus_CDS_NE_cds.txt",
        f"{ANNOTATE_SUBDIRS['poison_exon']}/{PREFIX}_nmd_sj_parsed.txt",

        # --- Step 04: ORF protein domain analysis ---
        f"{PFAMSCAN_OUT_DIR}/merged/{PREFIX}.pfam.txt",
        f"{INTERPROSCAN_OUT_DIR}/merged/{PREFIX}.tsv",
        f"{INTERPROSCAN_OUT_DIR}/merged/{PREFIX}.gff3",

        # --- Step 05: Overlaps with genomic elements ---
        f"05_genomic_element_overlaps/{PREFIX}_genomic_element_overlaps.txt",
        f"05_genomic_element_overlaps/{PREFIX}_SV_overlaps.txt",
      
        # --- Step 06: Overlaps with genomic elements ---
        f"06_tracks/{PREFIX}_reference_reclocus_CDS_extra.gtf",

        # --- Step 07: Get unique splice chains ---
        f"07_splicechains/{PREFIX}_splicechains.txt",

        # --- Step 08: Basic DEG analysis for brain-expressed genes ---
        f"08_DEG/{PREFIX}_limma.checkpoint",


        # f"09_collapsed-ORFs/{PREFIX}_reference_reclocus_CDS_aa_clust_header_generic.faa",
        # f"10_isoformswitchanalyzer_inputs/{PREFIX}_exp_isoforms-nt.fasta",
        # f"11_PoGo/{PREFIX}_PoGo_mm1_1MM.bed",


