#!/usr/bin/env snakemake

# ───────────────────────────────────────────────
# IMPORT 
# ───────────────────────────────────────────────

import pandas as pd
import os
import sys

from collections import defaultdict

# ───────────────────────────────────────────────
# CONFIG 
# ───────────────────────────────────────────────

# Define minimum snakemake version
min_snakemake_version = "8.25.0"

# Set snakemake main workdir variable. 
SNAKEDIR = workflow.basedir + '/'

# Load configuration from config file
configfile: SNAKEDIR + "../config/config.yaml"


# ───────────────────────────────────────────────
# LOCAL VARIABLE DEFINITIONS 
# ───────────────────────────────────────────────

# Utility function to make sure that values we expect to be booleans are passed as such
def as_bool(val):
    if isinstance(val, bool):
        return val
    if isinstance(val, str):
        return val.strip().lower() in {"true", "yes", "1"}
    return False


# ───────────────────────────────────────────────
# Define Output Directory Structure
# ───────────────────────────────────────────────

PREFIX = config["prefix"]

# Transdecoder outputs
TRANSDECODER_OUT_DIR = "02_ORF_prediction/transdecoder"
TRANSDECODER_LOG_DIR = f"logs/{TRANSDECODER_OUT_DIR}"

# GMST outputs
GMST_OUT_DIR = "02_ORF_prediction/gmst"
GMST_LOG_DIR = f"logs/{GMST_OUT_DIR}"

# Annotate output folders
ANNOTATE_DIR = "03_annotate"
ANNOTATE_SUBDIRS = {
    "annot":       f"{ANNOTATE_DIR}/A_annotation",
    "reclocus":    f"{ANNOTATE_DIR}/B_reclocus",
    "cds":         f"{ANNOTATE_DIR}/C_cds_integration",
    "asef":        f"{ANNOTATE_DIR}/D_ASEF",
    "final":       f"{ANNOTATE_DIR}/E_final",
    "poison_exon": f"{ANNOTATE_DIR}/F_poison_exons"
}


# ───────────────────────────────────────────────
# RULES
# ───────────────────────────────────────────────

include: "rules/01_isoform-counts.smk"
include: "rules/02_ORF_prediction_cpatv3.smk"
include: "rules/02_ORF_prediction_transdecoder.smk"
include: "rules/02_ORF_prediction_gmst.smk"
include: "rules/03_annotate.smk"


rule all:
    input:
        # --- Step 01: Isoform Counts ---
        f"01_isoform_counts/{PREFIX}_isoform-counts.txt",

        # --- Step 02: ORF Prediction ---
        # CPAT outputs
        f"02_ORF_prediction/cpat_prob/{PREFIX}_corrected.cpatv3p18.ORF_seqs.fa",
        f"02_ORF_prediction/cpat_leng/{PREFIX}_corrected.cpatv3l18.ORF_seqs.fa",
        # TransDecoder outputs
        f"{TRANSDECODER_OUT_DIR}/merged/{PREFIX}.transdecoder.pep",
        f"{TRANSDECODER_OUT_DIR}/merged/{PREFIX}.transdecoder.cds",
        f"{TRANSDECODER_OUT_DIR}/merged/{PREFIX}.transdecoder.gff3",
        f"{TRANSDECODER_OUT_DIR}/merged/{PREFIX}.transdecoder.bed",
        # GMST outputs
        f"{GMST_OUT_DIR}/merged/{PREFIX}.gmst.faa",
        f"{GMST_OUT_DIR}/merged/{PREFIX}.gmst.fnn",
        f"{GMST_OUT_DIR}/merged/{PREFIX}.gmst.gff3",

        # --- Step 03: Annotation & Analysis ---
        f"{ANNOTATE_SUBDIRS['final']}/{PREFIX}_reference_reclocus_refined.txt",
        f"{ANNOTATE_SUBDIRS['final']}/{PREFIX}_reference_reclocus_CDS_aa.fa",
        f"{ANNOTATE_SUBDIRS['asef']}/{PREFIX}_reference_reclocus_CDS_NE_cds.txt",
        f"{ANNOTATE_SUBDIRS['poison_exon']}/{PREFIX}_nmd_sj_parsed.txt",

      # expand("08_nmd_poison_exons/{prefix}_reference_reclocus_CDS_nmd_sj.txt", prefix=config["prefix"]),
      # expand("05_interpro/{prefix}_jobs-submitted.check", prefix=config["prefix"]),
      # expand("06_pfam/{prefix}_jobs-submitted.check", prefix=config["prefix"]),

      # expand("09_genomic_element_overlaps/{prefix}_genomic_element_overlaps.txt", prefix=config["prefix"]),
      # expand("09_genomic_element_overlaps/{prefix}_SV_overlaps.txt", prefix=config["prefix"]),
      # expand("10_tracks/{prefix}_reference_reclocus_CDS_extra.gtf", prefix=config["prefix"]),
      # expand("11_splicechains/{prefix}_patched_splicechains.txt", prefix=config["prefix"]),
      # expand("12_DEG/{prefix}_limma.checkpoint", prefix=config["prefix"]),
      # expand("13_collapsed-ORFs/{prefix}_reference_reclocus_CDS_aa_clust_header_generic.faa", prefix=config["prefix"]),
      # expand("14_isoformswitchanalyzer_inputs/{prefix}_exp_isoforms-nt.fasta", prefix=config["prefix"]),
      # expand("01_sqanti3/{prefix}_classification.txt", prefix=config["prefix"]),

      # final_output
