#!/usr/bin/env snakemake

# ───────────────────────────────────────────────
# IMPORT 
# ───────────────────────────────────────────────

import pandas as pd
import os
import sys

from collections import defaultdict

# ───────────────────────────────────────────────
# CONFIG 
# ───────────────────────────────────────────────

# Define minimum snakemake version
min_snakemake_version = "8.25.0"

# Set snakemake main workdir variable. 
SNAKEDIR = os.path.dirname(workflow.snakefile) + "/"

# Load configuration from config file
configfile: SNAKEDIR + "../config/config.yaml"


# ───────────────────────────────────────────────
# LOCAL VARIABLE DEFINITIONS 
# ───────────────────────────────────────────────

# Utility function to make sure that values we expect to be booleans are passed as such
def as_bool(val):
    if isinstance(val, bool):
        return val
    if isinstance(val, str):
        return val.strip().lower() in {"true", "yes", "1"}
    return False


# Set path variables to pass to bash scripts
shell_path       = expand("{util}:{ngs}:{igb}:{iso}:{lsf}:{sq3}/cDNA_Cupcake/sequence/:{sq3}cDNA_Cupcake/annotation/", 
                          util=config["utility_path"],  ngs=config["ngs_tools_path"],   igb=config["track_tools_path"],  
                          iso=config["isoseq_pipeline_path"],  sq3=config["sqanti_path"], lsf=config["minerva_queue_lsf_path"])
shell_maptools   = expand("{maptools}/NIAP/script/:{maptools}/ASEF/script/:{maptools}/poison_exon/script/:{maptools}/NIAP_classification_conversion/script/:{maptools}/CPAT_conversion/script/:{maptools}/reconstructed_locus/script/:{maptools}/CDS_prediction/script/",
                          maptools=config["isoseq_map_analysis_path"])
shell_perl5lib   = config["utility_perl5lib"]


# ───────────────────────────────────────────────
# RULES
# ───────────────────────────────────────────────

include: "rules/01_isoform-counts.smk"
#include: "rules/02_ORF_prediction.smk"
#include: "rules/03_interpro.smk"
#include: "rules/04_pfam.smk"
#include: "rules/05_annotate.smk"
#include: "rules/06_nmd_poison_exons.smk"
#include: "rules/07_genomic_element_overlaps.smk"
#include: "rules/08_tracks.smk"
#include: "rules/09_splicechains.smk"
#include: "rules/10_DEG.smk"
#include: "rules/11_collapsed-ORFs.smk"
#include: "rules/12_isoformswitchanalyzer_inputs.smk"
#include: "rules/13_pogo.smk"
#include: "rules/14_squanti3.smk"

rule all:
   input: 
      expand("01_isoform_counts/{prefix}_isoform-counts.txt", prefix=config["prefix"]),

      # expand("03_cpatv3/{prefix}_corrected.cpatv3p18.ORF_prob.tsv", prefix=config["prefix"]),
      # expand("04_transdecoder/{prefix}_transdecoder.check", prefix=config["prefix"]),
      # expand("05_interpro/{prefix}_jobs-submitted.check", prefix=config["prefix"]),
      # expand("06_pfam/{prefix}_jobs-submitted.check", prefix=config["prefix"]),

      # expand("07_annotate/{prefix}_reference.gtf", prefix=config["prefix"]),
      # expand("07_annotate/{prefix}_reference_transcript.txt", prefix=config["prefix"]),
      # expand("07_annotate/{prefix}_reference_gene.txt", prefix=config["prefix"]),
      # expand("08_nmd_poison_exons/{prefix}_reference_reclocus_CDS_nmd_sj.txt", prefix=config["prefix"]),

      # expand("09_genomic_element_overlaps/{prefix}_genomic_element_overlaps.txt", prefix=config["prefix"]),
      # expand("09_genomic_element_overlaps/{prefix}_SV_overlaps.txt", prefix=config["prefix"]),


      # expand("10_tracks/{prefix}_reference_reclocus_CDS_extra.gtf", prefix=config["prefix"]),
      # expand("11_splicechains/{prefix}_patched_splicechains.txt", prefix=config["prefix"]),
      # expand("12_DEG/{prefix}_limma.checkpoint", prefix=config["prefix"]),
      # expand("13_collapsed-ORFs/{prefix}_reference_reclocus_CDS_aa_clust_header_generic.faa", prefix=config["prefix"]),
      # expand("14_isoformswitchanalyzer_inputs/{prefix}_exp_isoforms-nt.fasta", prefix=config["prefix"]),
      # expand("01_sqanti3/{prefix}_classification.txt", prefix=config["prefix"]),

      # final_output
